[WOR-2124] Adjust docs for interrupting boundary event (#9066)
mendix/docs: content/en/docs/refguide/modeling/application-logic/workflows/boundary-events.md

---

{{% alert color="info" %}}
Non-interrupting timer boundary events were introduced in beta in Studio Pro 10.15 and released in GA in Studio Pro 10.16.

> Interrupting timer boundary events were introduced in Studio Pro 10.20 and is currently in beta. Certain features or attributes are subject to change and may contain bugs. For more information, see [Beta and Experimental Releases](/releasenotes/beta-features/).

Reviewer: "is" => "are" in beta

There are two main types of boundary events:

> * Non-interrupting boundary events: These events do not interrupt the ongoing activity. When triggered, they allow the activity to continue while simultaneously starting a new path from the boundary event. Use non-interrupting boundary events when the parent activity should remain active, but you would like to do something in addition to the active activity/path. For example, after 2 days, a reminder should be sent to the assigned user. As per BPMN 2.0 specification, non-interrupting boundary events are visualized as two dashed circles with an icon in the center.

Reviewer 2: but you would like to do something in addition to the active activity/path. => Maybe something like: but you would like to do something in parallel?

There are two main types of boundary events:

* Non-interrupting boundary events: These events do not interrupt the ongoing activity. When triggered, they allow the activity to continue while simultaneously starting a new path from the boundary event. As per BPMN 2.0 specification, non-interrupting boundary events are visualized as two dashed circles with an icon in the center.

> * Interrupting boundary events (currently in beta): When these events are triggered, they interrupt the normal path of the activity they are attached to. The activity stops and the process flow is redirected to the boundary event's outgoing sequence path. As per BPMN 2.0 specification, interrupting boundary events are visualized as two solid circles.

Reviewer 3: ```suggestion
* Interrupting boundary events (currently in beta): When these events are triggered, they interrupt the activity they are attached to, meaning that the activity will be aborted. The process flow is redirected to the boundary event's outgoing sequence path. As per BPMN 2.0 specification, interrupting boundary events are visualized as two solid circles.
```

Boundary Events are always displayed by 2 circles (either solid or dashed) and are linked by a dotted line to the parent activity. The icon inside the event indicates the type of event. For example, a clock indicates that it is a timer boundary event.

> Below is an example of how a non-interrupting timer boundary event looks like:

Reviewer: "how" => "what"

* Non-interrupting boundary events: These events do not interrupt the ongoing activity. When triggered, they allow the activity to continue while simultaneously starting a new path from the boundary event. As per BPMN 2.0 specification, non-interrupting boundary events are visualized as two dashed circles with an icon in the center.

> * Interrupting boundary events: When these events are triggered, they interrupt the normal path of the activity they are attached to. The activity stops and the process flow is redirected to the boundary event's outgoing sequence path. As per BPMN 2.0 specification, interrupting boundary events are visualized as two solid circles. 

Reviewer 4: suggesting to remove "the normal path of" as it interrupt the activity. The the path is interrupted, is a consequence of that.

change "...activity stops..." to "...activity is aborted ..." (abort is a defined state, stop is not).

Reviewer 5: These comments are resolved in the latest revision:

> Interrupting boundary events (currently in beta): When these events are triggered, they interrupt the activity they are attached to, meaning that this activity will be aborted. The process flow is redirected to the boundary event's outgoing sequence path. As per BPMN 2.0 specification, interrupting boundary events are visualized as two solid circles.

> * Interrupting boundary events: When these events are triggered, they interrupt the normal path of the activity they are attached to. The activity stops and the process flow is redirected to the boundary event's outgoing sequence path. As per BPMN 2.0 specification, interrupting boundary events are visualized as two solid circles. 

Reviewer 4: When to use non-interrupting and when to use interrupting boundary event.
Use non-interrupting boundary events when the parent activity should remain active, but you would like to do someth9ng in addition to the active activity/path. For example after 2 days a reminder should be send to the assigned user.
Use interrupting boundary event in situations where further execution of the activity (and other following activities), in not required/desired and an alternative path should be taken. For example to start an escalation or "fast-track" path when an activity is not completed 2 days after due date. Or when the assigned user does not make a decision within 5 days, the user task is aborted and the process will continue with a pre-set decision.

Reviewer 5: Added these to the introduction section: https://github.com/mendix/docs/pull/9066/commits/f6d1b0e8c27643961a9bcabbd8e4f771199fe4d9.

    {{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/edit-dialog.png" alt="Adding boundary events through edit dialog" width="450" >}}

To configure the properties of a boundary event, double-click the event to open its properties dialog box. For more information on how to configure the properties of a timer boundary event, see the [Properties](/refguide/timer/#properties) section in *Timer*.

#### Enabling and Adding Interrupting Boundary Events (Beta)

Interrupting boundary events are currently in beta. To enable this feature, go to Studio Pro **Preferences** -> **New features** -> **Workflow editor** > **Enable interrupting timer boundary events (beta)**:

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/enable-interrupting-boundary-event.png" alt="Enable interrupting boundary event" width="450">}}

After adding interrupting boundary events is enabled, the **Boundary properties** section shows up in the properties dialog box of a boundary event.

> By default, the **Interrupting** property is set to **No**, which means that it is by default adding a non-interrupting boundary event. To add an interrupting boundary event, set the **Interrupting** property to **Yes**. Click **OK** to save your configuration:

Reviewer 3: ```suggestion
By default, the **Interrupting** property is set to **No**, which means that the initially added event is a non-interrupting boundary event. To add an interrupting boundary event, set the **Interrupting** property to **Yes**. Click **OK** to save your configuration:
```

Reviewer 3: Also, would this be good location to inform about the risks of switching between the two events?

* Double-click one of the above-listed activities to open its properties dialog box, go to the **Events** tab, and in the > **Boundary events** section, click **Add timer event** to add a timer boundary event:

    {{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/edit-dialog.png" alt="Adding boundary events through edit dialog" width="450" >}}

To configure the properties of a boundary event, double-click the event to open its properties dialog box. For more information on how to configure the properties of a timer boundary event, see the [Properties](/refguide/timer/#properties) section in *Timer*.

#### Enabling and Adding Interrupting Boundary Events (Beta)

Interrupting boundary events are currently in beta. To enable this feature, go to Studio Pro **Preferences** -> **New features** -> **Workflow editor** > **Enable interrupting timer boundary events (beta)**:

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/enable-interrupting-boundary-event.png" alt="Enable interrupting boundary event" width="450">}}

> After adding interrupting boundary events is enabled, the **Boundary properties** section shows up in the properties dialog box of a boundary event.

Reviewer: I would remove the word "adding":

> After interrupting boundary events are enabled

By default, the **Interrupting** property is set to **No**, which means that the initially added event is a non-interrupting boundary event. To add an interrupting boundary event, set the **Interrupting** property to **Yes**. Click **OK** to save your configuration:

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/interrupting-property.png" alt="Changing the type of boundary event" width="450">}}

#### Implications of changing boundary event type

When you change a boundary event from non-interrupting to interrupting or vice versa, you will be presented with the following dialog.

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/security-dialog.png" alt="Security Dialog when changing type" width="450">}}

> The boundary event will be re-created as the specified type meaning that it will also generate new Id's. The application runtime will use these new Id's to verify what part in the workflow has changed and how to resolve the conflicts.

Reviewer 3: ```suggestion
The boundary event will be re-created as the specified type. As a result, new IDs will be created. These IDs are used by the application's runtime for its conflict detection analysis.
```

> This will mean that active workflows with scheduled boundary events will become incompatible and need to be restarted.

Reviewer 3: ```suggestion
This will mean that active workflows with scheduled boundary events will become incompatible and need to be restarted.
```
```suggestion
Any scheduled boundary event which has been toggled will become incompatible and needs to be restarted.
```

After interrupting boundary events are enabled, the **Boundary properties** section is displayed in the properties dialog box of a boundary event.

By default, the **Interrupting** property is set to **No**, which means that the initially added event is a non-interrupting boundary event. To add an interrupting boundary event, set the **Interrupting** property to **Yes**. Click **OK** to save your configuration:

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/interrupting-property.png" alt="Changing the type of boundary event" width="450">}}

#### Implications of Changing the Boundary Event Type

For an existing boundary event, when you change its type from non-interrupting to interrupting or vice versa, you will be presented with a warning dialog. For example, when you change a boundary event from non-interrupting to interrupting, you will see the following warning dialog:

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/security-dialog.png" alt="Security Dialog when changing type" width="450">}}

After you confirm the change:

> * The boundary event is re-created as the specified type. As a result, new IDs are created. These IDs are used by the application's runtime for its conflict detection analysis.

Reviewer 5: @timvandervoorn what does "the application's runtime" mean here? Does it mean "the Mendix Runtime"?

Reviewer 2: Yes the mendix runtime

By default, the **Interrupting** property is set to **No**, which means that the initially added event is a non-interrupting boundary event. To add an interrupting boundary event, set the **Interrupting** property to **Yes**. Click **OK** to save your configuration:

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/interrupting-property.png" alt="Changing the type of boundary event" width="450">}}

#### Implications of Changing the Boundary Event Type

For an existing boundary event, when you change its type from non-interrupting to interrupting or vice versa, you will be presented with a warning dialog. For example, when you change a boundary event from non-interrupting to interrupting, you will see the following warning dialog:

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/boundary-events/security-dialog.png" alt="Security Dialog when changing type" width="450">}}

After you confirm the change:

* The boundary event is re-created as the specified type. As a result, new IDs are created. These IDs are used by the application's runtime for its conflict detection analysis.

> * Any scheduled boundary event which has been toggled becomes incompatible and needs to be restarted.

Reviewer 2: Maybe we can add a little bit more information like: 

The subsequent flow of the scheduled boundary event which has been toggled becomes incompatible and needs to be restarted.

Reviewer 5: Updated to: 

"The current activities on the boundary event path including the boundary event itself become incompatible and need to be restarted."

    {{% alert color="info" %}}This does not change the order execution of the paths, as this is dependent on the expiration of the timer.{{% /alert %}}

### Adding interrupting boundary events

Interrupting boundary events can be added by going to the properties of a non-interrupting boundary event. Within the properties of a boundary event is a radio button which sets either the boundary event to interrupting or non-interrupting. The type change however is only fullfilled when the OK button is pressed.

> 

Reviewer 4: I've noticed when doing this the first time there is a warning that the change from non-interrupting to interrupting will recreated the path. PLs mention this hear and maybe also an explanation why the recreate is necessary.

Reviewer 5: A whole section is added for this: See the latest version here: https://github.com/mendix/docs/pull/9066/files.

> 

Reviewer 4: A maximum of 5 non-interrupting boundary events can be added to a single activity. This is mainly a visualization limitation and no technical limitation. There can only be 1 interrupting boundary event per activity.

Reviewer 5: Added as an info alert at the beginning of the **Enabling and Adding Interrupting Boundary Events (Beta)** section.

{{< figure src="/attachments/refguide/modeling/application-logic/workflows/workflow-elements/boundary-events/radio-button.png" alt="Changing the type of boundary event" width="450" class="no-border">}}

## Execution

Boundary events are initiated when their parent activity is initiated. For example, for a timer with a fixed duration, it will start its count down when the parent activity is initiated. When the parent activity is completed before any of the boundary events are triggered, none of the activities in the boundary event path will be executed and all timers will be cancelled.

### Non-interrupting boundary events
With non-interrupting boundary events the parent activity remains active/in progress when an event is triggered (i.e., the parent activity is not interrupted). For example, when a timer event on a user task is triggered after 2 days, this task will remain in progress and the path defined below the timer event is executed. When the boundary event path reaches the **End of Boundary Path** activity, the workflow will await the completion of the parent activity.

### Interrupting boundary events

> However with interrupting boundary events, the parent activity as the name implies is interrupted. For example, when an interrupting timer boundary event is set on a user task and is also triggered after 2 days, this task will be cancelled, and the path defined below the timer event is now the only path which will be finished. Currently only **End Workflow** activities are allowed at the end of an interrupting boundary event. 

Reviewer 4: "...the parent activity is **aborted**".

"...path defined below the timer event ..." this depends on the direction of the process. Suggestion: "... path following/from the boundary event is now the active path.

" Currently only **End Workflow** activities are allowed at the end of an interrupting boundary event." Can we rewrite and mention this is "due to" being Beta and that a Jump from the interrupting path back to the path of the parent activity will be added soon.

Reviewer 5: All comments are resolved in commit: https://github.com/mendix/docs/pull/9066/commits/f6d1b0e8c27643961a9bcabbd8e4f771199fe4d9.

Boundary events are initiated when their parent activity is initiated. For example, for a timer with a fixed duration, it will start its count down when the parent activity is initiated. When the parent activity is completed before any of the boundary events are triggered, none of the activities in the boundary event path will be executed and all timers will be cancelled.

### Non-Interrupting Boundary Events

With non-interrupting boundary events the parent activity remains active/in progress when an event is triggered (i.e., the parent activity is not interrupted). For example, when a timer event on a user task is triggered after 2 days, this task will remain in progress and the path defined below the timer event is executed. When the boundary event path reaches the **End of Boundary Path** activity, the workflow will await the completion of the parent activity.

### Interrupting Boundary Events

> However with interrupting boundary events, the parent activity as the name implies is interrupted. For example, when an interrupting timer boundary event is set on a user task and is also triggered after 2 days, this task will be cancelled, and the path defined below the timer event is now the only path which will be finished. Currently only **End Workflow** activities are allowed at the end of an interrupting boundary event. 

Reviewer 3: ```suggestion
However with interrupting boundary events, the parent activity is interrupted. For example, when an interrupting timer boundary event is set on a user task and is triggered after 2 days, this task will be aborted, and the path defined below the timer event is now the active path. Currently only **End Workflow** activities are allowed at the end of an interrupting boundary event. 
```

Reviewer 3: What about certain cases like other active non interrupting boundary flows. These will get aborted, right? Please verify and check if we should describe such behavior here.

Reviewer 3: We also need to mention that currently, the interrupting boundary event path has to end with an `End` activity. This, subsequently, means that the workflow will be ended after the flow has reached it end.

Reviewer 5: Dave resolved those comments in commit: https://github.com/mendix/docs/pull/9066/commits/595528db4cee3cebf35764966e1062b1d780d18b.

Reviewer 5: I've added an info alert based on [Paul's feedback below](https://github.com/mendix/docs/pull/9066#discussion_r1967264324). See below the info alert:

{{% alert color="info" %}}
Currently, due to its beta status, an interrupting boundary event must end with an **End** event. The option to allow jumping back to the path of the parent activity from the interrupting path will be added soon. For more limitations, see the [Current Limitations](#limitation) section below.
{{% /alert %}}

Boundary events are initiated when their parent activity is initiated. For example, for a timer with a fixed duration, it will start its count down when the parent activity is initiated. When the parent activity is completed before any of the boundary events are triggered, none of the activities in the boundary event path will be executed and all timers will be cancelled.

### Non-Interrupting Boundary Events

With non-interrupting boundary events the parent activity remains active/in progress when an event is triggered (i.e., the parent activity is not interrupted). For example, when a timer event on a user task is triggered after 2 days, this task will remain in progress and the path defined below the timer event is executed. When the boundary event path reaches the **End of Boundary Path** activity, the workflow will await the completion of the parent activity.

### Interrupting Boundary Events

> However with interrupting boundary events, the parent activity is interrupted. For example, when an interrupting timer boundary event is set on a user task and is triggered after 2 days, this task will be aborted, and the path defined below the timer event is now the active path. Currently only **End Workflow** activities are allowed at the end of an interrupting boundary event. Meaning that at the end of an interrupting boundary event the workflow will be ended.

Reviewer 3: ```suggestion
However with interrupting boundary events, the parent activity is interrupted. For example, when an interrupting timer boundary event is set on a user task and is triggered after 2 days, this task will be aborted, and the path defined below the timer event will become the active path. Currently only **End Workflow** activities are allowed at the end of an interrupting boundary event.
```

{{% alert color="info" %}}

> When there are non-interrupting boundary events specified for the parent activity, and the interrupting boundary event is executed. It will also cancel all the scheduled and running non-interrupting boundary events.{{% /alert %}}

Reviewer 3: ```suggestion
When there are non-interrupting boundary events specified for the parent activity, and the interrupting boundary event is executed, all the scheduled and running non-interrupting boundary events will be aborted.{{% /alert %}}
```

* `$ParentTask` – the parent user task of the attached boundary event
* `$CalledWorkflowInstance` – the parent Call workflow activity of the attached boundary event

## Current Limitations

The current release of boundary events has the following limitations which is actively being developed:

* Interrupting boundary events, are not allowed to be nested.

> * Interrupting boundary events are not allowed to have an jump at the end of their flow, nor do they allow jumping outside of their flow.

Reviewer 4: Also jumping in a boundary path is not possible (yet)

Reviewer 5: Added also in commit: https://github.com/mendix/docs/pull/9066/commits/f6d1b0e8c27643961a9bcabbd8e4f771199fe4d9.

* `$ParentTask` – the parent user task of the attached boundary event
* `$CalledWorkflowInstance` – the parent Call workflow activity of the attached boundary event

## Current Limitations

The current release of boundary events has the following limitations which is actively being developed:

> * Interrupting boundary events, are not allowed to be nested.

Reviewer 3: ```suggestion
* Interrupting boundary events are not allowed to be nested in other boundary event paths, meaning that if an activity has a boundary event path (interrupting or non-interrupting), this activity cannot be placed in a boundary event path.
```

* Interrupting boundary events are not allowed to have an jump at the end of their flow, nor do they allow jumping outside of their flow.

> * Non-interrupting timer boundary events currently have no recurrence (they are only executed once and will not repeat).

Reviewer 3: Is this a limitation or just a feature we haven't released yet. I would suggest to not mention this.

Author: Recurrence was mentioned before as a limitation. So I can indeed remove this but given it was already there. Maybe we can keep it in.

Reviewer 3: If it was already there, then let's keep it.

* `$ParentTask` – the parent user task of the attached boundary event
* `$CalledWorkflowInstance` – the parent Call workflow activity of the attached boundary event

## Current Limitations

> The current release of boundary events has the following limitations which is actively being developed:

Reviewer: > The current release of boundary events is still under development and has the following limitations:

* Interrupting boundary events are not allowed to be nested in other boundary event paths, meaning that if an activity has a boundary event path (interrupting or non-interrupting), this activity cannot be placed in a boundary event path.

> * Interrupting boundary events are not allowed to have an jump at the end of their flow, nor do they allow jumping outside of their flow.

Reviewer: "an" => "a" jump
